core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala

```diff
@@ -608,31 +617,196 @@ trait CDCReaderImpl extends DeltaLogging {
+   * however, is to compare two DVs and apply the diff to the file to get removed and re-added rows.
+   *
+   * Currently it is always the case that in the log "remove" comes first, followed by "add" --
+   * which means that the file stays alive with a new DV. There's another possibility, though not
+   * make many senses, that a file is "added" to log then "removed" in the same version. If this
+   * becomes possible in future, we have to reconstruct the timeline considering the order of
+   * actions rather than simply matching files by path.
+   */
   protected def getDeletedAndAddedRows(
       addFileSpecs: Seq[CDCDataSpec[AddFile]],
       removeFileSpecs: Seq[CDCDataSpec[RemoveFile]],
       deltaLog: DeltaLog,
       snapshot: Snapshot,
       isStreaming: Boolean,
       spark: SparkSession): Seq[DataFrame] = {
-    val dfs = ListBuffer[DataFrame]()
+    // Transform inputs to maps indexed by version and path and map each version to a CommitInfo
+    // object.
+    val versionToCommitInfo = MutableMap.empty[Long, CommitInfo]
+    val addFilesMap = addFileSpecs.flatMap { spec =>
+      spec.commitInfo.ifDefined { ci => versionToCommitInfo(spec.version) = ci }
+      spec.actions.map { action =>
+        val key =
+          FileVersionWithPath(spec.version, spec.timestamp, action.path, spec.commitInfo)
+        key -> action
+      }
+    }.toMap
+    val removeFilesMap = removeFileSpecs.flatMap { spec =>
+      spec.commitInfo.ifDefined { ci => versionToCommitInfo(spec.version) = ci }
+      spec.actions.map { action =>
+        val key =
+          FileVersionWithPath(spec.version, spec.timestamp, action.path, spec.commitInfo)
+        key -> action
+      }
+    }.toMap
 
-    if (addFileSpecs.nonEmpty) {
-      dfs.append(scanIndex(
-        spark,
-        new CdcAddFileIndex(spark, addFileSpecs.toSeq, deltaLog, deltaLog.dataPath, snapshot),
-        snapshot.protocol,
-        snapshot.metadata,
-        isStreaming))
+    val finalAddFiles = MutableMap[FileVersion, ListBuffer[AddFile]]()
+    val finalRemoveFiles = MutableMap[FileVersion, ListBuffer[RemoveFile]]()
+
+    // If a path is only being added, then scan it normally as inserted rows
+    (addFilesMap.keySet -- removeFilesMap.keySet).foreach { addKey =>
+      finalAddFiles
+        .getOrElseUpdate(new FileVersion(addKey), ListBuffer())
+        .append(addFilesMap(addKey))
     }
-    if (removeFileSpecs.nonEmpty) {
-      dfs.append(scanIndex(
-        spark,
-        new TahoeRemoveFileIndex(
-          spark, removeFileSpecs.toSeq, deltaLog, deltaLog.dataPath, snapshot),
-        snapshot.protocol,
-        snapshot.metadata,
-        isStreaming))
+
+    // If a path is only being removed, then scan it normally as removed rows
+    (removeFilesMap.keySet -- addFilesMap.keySet).foreach { removeKey =>
+      finalRemoveFiles
+        .getOrElseUpdate(new FileVersion(removeKey), ListBuffer())
```

Comment:
`FileVersion` seems like the wrong name? It looks like it is referring the the delta log version?